New rotate docs

DanielMurphy22 opened the pull request:
**Description of work.**
- Added new rotation examples to HowToDefineGeometricShape. Added new plotting examples to MeshPlots. SetSample and CopySample have both been updated already with references to SetGoniometer, but an indentation error in SetSample has been fixed here.
- A section for adding arrows has be added to show how to add the beam arrow and how to add the crystal axes.

**To test:**
- build docs-html target
- check the links and new examples MeshPlotHelp page and HowToDefineGeometricShape page. Check that the plot for each example appears. Note that for some examples the actual plotting script is hidden, and a code-block instead shows the user the important commands.The full script for each plot should be available via `(Source code, png, hires.png, pdf)` above the plot image. Note the MeshObject Plot (which is not a new example) may not work when you build the docs locally as it may not find tube.stl, but it works for the [current online version](https://docs.mantidproject.org/nightly/plotting/MeshPlotHelp.html#id3).
- Also there is a fix in this PR to a example code recently added to SetSample. Check the plot there now appears correctly.

fixes #32031

*This does not require release notes* because **this is documentation**


---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DanielMurphy22 committed:
Update PlotMesh and GeometricShape Docs

Add rotating tags to HowtoDefineGeometricShapes
Add CSGObjects to MeshPlotHelp
Also add an arrow to point from source to sample

DanielMurphy22 committed:
Merge branch 'master' into New_Rotate_docs

DanielMurphy22 committed:
Add arrow method and example for plotting crystal axes on Mesh Plot

DannyHindson commented on docs/source/concepts/HowToDefineGeometricShape.rst:
```diff
@@ -340,6 +340,54 @@ axis runs from the start aperture to the end aperture. "Height" is along
+All these rotatable shapes (expect for `Infinite Plane`_ and `Infinite Cylinder`_) can be plotted to check
+your shape definition is correct. For more details see :ref:`Mesh_Plots`.
+
+Shapes will be automatically rotated, if a rotation is set using :ref:<algm-SetGoniometer>.
```
I think a short mention of the fact that you can still use the rotate and rotate-all tags in combination with the automatic goniometer rotation would be useful

DannyHindson commented on docs/source/plotting/MeshPlotHelp.rst:
```diff
@@ -31,12 +31,17 @@ Mesh Plots
 **Mesh Plots can only be accessed with a script, not through the Workbench interface**
+Here the mesh is plotted as a Poly3DCollection `Polygon <https://matplotlib.org/stable/tutorials/toolkits/mplot3d.html#polygon-plots>`_.
+
+These sample shapes can be created with :ref:`algm-SetSample` and copied using :ref:`algm-CopySample`.
```
I think LoadSampleShape and LoadSampleEnvironment should also be mentioned here as ways of setting up a mesh sample shape

DannyHindson commented on docs/source/plotting/MeshPlotHelp.rst:
```diff
@@ -76,7 +81,294 @@ Basic example of plotting a general Poly3DCollection `Polygon <https://matplotli
+For more plotting advice: :ref:`02_scripting_plots`
 
 |
 |
```
Is this content from the "Plot Toolbar" heading onwards required - I wasn't clear if this was a duplicate of the help for the general plot window or whether it is tailored to mesh plotting. The content under the General heading near the end where it describes the "Plots Toolbox" and the "File Settings" window appears to be the same as the normal plot window.

If it is tailored to mesh plotting then it's fine to leave in but perhaps a sentence after the "Plot Toolbar" heading to explain that the mesh plotting provides a reduced set of toolbar options would be good

DanielMurphy22 replied:
I think you're right, there's a lot here that isn't specific to MeshPlots and the other documentation on plotting and 3D plots already supports this.

DannyHindson commented on docs/source/plotting/MeshPlotHelp.rst:
```diff
@@ -76,7 +81,294 @@ Basic example of plotting a general Poly3DCollection `Polygon <https://matplotli
+    )
+
+   '''Add arrow along beam direction'''
+   source = ws.getInstrument().getSource().getPos()
```
Similar comment to higher up. At the moment this script doesn't have a line to create ws. I'm guessing this is intentional to avoid repeating the python to do this but perhaps a single line comment saying "create ws as previously described" would be useful

DannyHindson commented on docs/source/plotting/MeshPlotHelp.rst:
```diff
@@ -76,7 +81,294 @@ Basic example of plotting a general Poly3DCollection `Polygon <https://matplotli
+For more plotting advice: :ref:`02_scripting_plots`
 
 |
 |
```
The figure shown under the Options menu heading shows two tabs: Axes and Images. I only see the Axes tab

DanielMurphy22 replied:
Great spot, I am removing this as it more general to all 3D plots

DannyHindson commented on docs/source/plotting/MeshPlotHelp.rst:
```diff
@@ -76,7 +81,294 @@ Basic example of plotting a general Poly3DCollection `Polygon <https://matplotli
+   # Component index 0 is the Container:
+   # container_mesh = environment.getComponent(0).getShape().getMesh()
+   component_mesh = environment.getComponent(2).getMesh()
+
```
This script doesn't plot anything at the moment. I'm guessing this is deliberate because it avoids repeating the plotting lines already outlined twice higher up. Perhaps a single line comment in the script saying "...plot as previously described" or similar would make this totally clear

DannyHindson reviewed (changes requested):
Must have not checked the plots on my review of the earlier PR! The SetSample documentation looks good now though

I've added a few bits and pieces but generally looks very good.

Also there's one small glitch in the table of contents for the "3D mesh plots..." help page where the filename for the gear icon .png appears rather than an "O" or the gear itself. This predates this PR though:

![Screenshot from 2021-08-09 10-52-33](https://user-images.githubusercontent.com/56295817/128689049-2d520cf4-1543-4cc9-b9af-ae321296f9bd.png)

DanielMurphy22 committed:
set_mesh_axes_equal() and requested changes

Added the set_mesh_axes_equal method to centre and setand equal aspect ratio on a mesh plot
This new method has been utilised in MeshPlotHelp, SetSample and CopySample
MeshPlotHelp has been simplified to remove general plotting advice
Useful comments ahve been added to code snippets
All other PlotHelpDocs have the Options Menu name corrected in the contents

DannyHindson reviewed (commented):
I've retested this and looked through the docs and all looks v good. I'll aim to approve on Fri when I'm back. I have two thoughts I wanted to share - both these are optional I think for this PR:

- Do we need a comment in the "Rotating shapes" section about the fact that some of the shapes already have an in built method of rotating themselves? eg cylinder already has an axis property. If I remember correctly we decided to introduce a general `rotate` tag in the end because some shapes were missing a native rotation property

- One of the drivers for this bit of work was WISH wanting to set up and rotate cuboid shapes. I've previously looked at the cuboid shape section on the "How to define Geometric shape page" and I'm not totally sure the xml matches up with the words. If you have a minute please could you check? It's the sentence(s) about where the xyz axis and planes are that didn't seem to me to correspond to the xml given. This bit isn't essential as part of this PR because you've not changed this

DanielMurphy22 committed:
Rewrite Cuboid Geometry examples with plt directives

Also the axis tag is now mentioned in the rotating shapes section

DannyHindson commented on docs/source/concepts/HowToDefineGeometricShape.rst:
```diff
@@ -340,6 +452,59 @@ axis runs from the start aperture to the end aperture. "Height" is along
+        <rotate x="90" y="-45" z="0" />
+      </cylinder>
+
+      <algebra val="stick" />
```
Given that the cylinder could be rotated as required using just the axis tag, I think it would be better to use the cuboid here. Please can you update to an example using the width\height\depth representation of a cuboid?

DannyHindson commented on docs/source/concepts/HowToDefineGeometricShape.rst:
```diff
@@ -340,6 +452,59 @@ axis runs from the start aperture to the end aperture. "Height" is along
+~~~~~~~~~~~~~~~
+
+*Note that some shapes (such as* `Cylinder (finite height) <Cylinder_>`_ *and* `Tapered Guide`_ *) can be oriented in a
+certain direction, using the* ``axis`` *tag, but this does not support all possible 3D rotations.*
```
I suggest you extend to say "... using the ``axis`` tag, but this representation isn't general enough to support all possible 3D rotations for all shapes". 
I think the axis tag does actually allow all possible 3D rotations for the cylinder. It is just for shapes that aren't symmetrical about a single axis where a single axis tag wasn't going to work

DannyHindson commented on docs/source/concepts/HowToDefineGeometricShape.rst:
```diff
@@ -340,6 +452,59 @@ axis runs from the start aperture to the end aperture. "Height" is along
+All these rotatable shapes (expect for `Infinite Plane`_ and `Infinite Cylinder`_) can be plotted to check
+your shape definition is correct. For more details see :ref:`Mesh_Plots`.
+
+Shapes will be automatically rotated, if a rotation is set using :ref:`algm-SetGoniometer`. This can work alongside
```
Do the goniometer rotations happen about the origin (same as rotate-all)? Worth adding a few words to say that here

DannyHindson reviewed (changes requested):
Thanks for the changes - the new cuboid section looks good. I've added a few small\slightly pedantic comments about the "rotating shapes" section and then I think we're done

DanielMurphy22 committed:
Changes to Rotating Shapes section

Use Cuboid rather than Cylinder in rotate tag example
Clarify axis and goniometer tags

DannyHindson reviewed (dismissed):
Looks good!
